File: 다이나믹프로그래밍/주지수_15724.java
package 다이나믹프로그래밍;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/*
4 4
9 14 29 7
1 31 6 13
21 26 40 16
8 38 11 23
3
1 1 3 2
1 1 1 4
1 1 4 4

2 3
10 9 8
1 2 3
3
1 1 2 2
1 1 2 3
2 2 2 3
 */
public class 주지수_15724 {
	static int N;
	static int M;
	static int[][] map;
	static int[][] dp;

	public static void main(String[] args) throws Exception {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	    StringTokenizer st;

		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		map = new int[N + 1][M + 1];
		dp = new int[N + 1][M + 1];

		for (int i = 1; i <= N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 1; j <= M; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		for (int r = 1; r <= N; r++) {
			dp[r][1] = dp[r - 1][1] + map[r][1];
		}
		for (int c = 1; c <= M; c++) {
			dp[1][c] = dp[1][c - 1] + map[1][c];
		}
		for (int i = 2; i <= N; i++) {
			for (int j = 2; j <= M; j++) {
				dp[i][j] = map[i][j] + dp[i][j - 1] + dp[i - 1][j] - dp[i - 1][j - 1];
			}
		}

		StringBuilder answer = new StringBuilder();

		int Q = Integer.parseInt(br.readLine());
		for (int i = 0; i < Q; i++) {
			st = new StringTokenizer(br.readLine());
			int r1 = Integer.parseInt(st.nextToken());
			int c1 = Integer.parseInt(st.nextToken());
			int r2 = Integer.parseInt(st.nextToken());
			int c2 = Integer.parseInt(st.nextToken());

			int value = dp[r2][c2] - dp[r2][c1 - 1] - dp[r1 - 1][c2] + dp[r1 - 1][c1 - 1];
			answer.append(value).append("\n");
		}

		System.out.print(answer);
	}
}

File: 코틀린_코테스터디/week2/주지수_15724.kt
package 코틀린_코테스터디.week2

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

/*
4 4
9 14 29 7
1 31 6 13
21 26 40 16
8 38 11 23
3
1 1 3 2
1 1 1 4
1 1 4 4
 */

private var N: Int = 0
private var M: Int = 0
private lateinit var origin: Array<Array<Int>>
private lateinit var dp: Array<Array<Int>>
private val br = BufferedReader(InputStreamReader(System.`in`))

fun main() {
    val st = StringTokenizer(br.readLine())

    N = st.nextToken().toInt()
    M = st.nextToken().toInt()

    origin = Array(N + 2) { Array(M + 1) { 0 } }
    dp = Array(N + 2) { Array(M + 1) { 0 } }

    repeat(N) { i ->
        val st = StringTokenizer(br.readLine())
        repeat(M) { j ->
            origin[i + 1][j + 1] = st.nextToken().toInt()
        }
    }
    initDP()
    calc()
}

private fun calc() {
    val k = br.readLine().toInt()
    val sb = StringBuilder()

    repeat(k) {
        val st = StringTokenizer(br.readLine())
        val y1 = st.nextToken().toInt()
        val x1 = st.nextToken().toInt()
        val y2 = st.nextToken().toInt()
        val x2 = st.nextToken().toInt()

        var area = dp[y2][x2] - dp[y1 - 1][x2] - dp[y2][x1 - 1] + dp[y1 - 1][x1 - 1]
        sb.append("$area\n")
    }
    print(sb)
}

private fun initDP() {
    dp[1][1] = origin[1][1]
    for (i in 1..N) {
        for (j in 1..M) {
            if (i == 1) {
                if (j == 1) continue
                dp[i][j] = dp[i][j - 1] + origin[i][j]
            } else if (j == 1) {
                dp[i][j] = dp[i - 1][j] + origin[i][j]
            } else {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1] - dp[i - 1][j - 1] + origin[i][j]
            }
        }
    }
}
